polygon/sync/canonical_chain_builder.go

```diff
@@ -0,0 +1,246 @@
+	Tip() *types.Header
+	HeadersInRange(start uint64, count uint64) []*types.Header
+	Prune(newRootNum uint64) error
+	Connect(headers []*types.Header) error
+}
+
+type producerSlotIndex uint64
+
+type forkTreeNode struct {
+	parent   *forkTreeNode
+	children map[producerSlotIndex]*forkTreeNode
+
+	header     *types.Header
+	headerHash libcommon.Hash
+
+	totalDifficulty uint64
+}
+
+type canonicalChainBuilderImpl struct {
+	root *forkTreeNode
+	tip  *forkTreeNode
+
+	difficultyCalc DifficultyCalculator
+}
+
+func NewCanonicalChainBuilder(
+	root *types.Header,
+	difficultyCalc DifficultyCalculator,
+) CanonicalChainBuilder {
+	impl := &canonicalChainBuilderImpl{
+		difficultyCalc: difficultyCalc,
+	}
+	impl.Reset(root)
+	return impl
+}
+
+func (impl *canonicalChainBuilderImpl) Reset(root *types.Header) {
+	impl.root = &forkTreeNode{
+		children:   make(map[producerSlotIndex]*forkTreeNode),
+		header:     root,
+		headerHash: root.Hash(),
+	}
+	impl.tip = impl.root
+}
+
+// depth-first search
+func (impl *canonicalChainBuilderImpl) enumerate(visitFunc func(*forkTreeNode) bool) {
+	stack := []*forkTreeNode{impl.root}
+	for len(stack) > 0 {
+		// pop
+		node := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		if !visitFunc(node) {
+			break
+		}
+
+		for _, child := range node.children {
+			stack = append(stack, child)
+		}
+	}
+}
+
+func (impl *canonicalChainBuilderImpl) nodeByHash(hash libcommon.Hash) *forkTreeNode {
+	var result *forkTreeNode
+	impl.enumerate(func(node *forkTreeNode) bool {
+		if node.headerHash == hash {
+			result = node
+		}
+		return result == nil
+	})
+	return result
+}
+
+func (impl *canonicalChainBuilderImpl) ContainsHash(hash libcommon.Hash) bool {
+	return impl.nodeByHash(hash) != nil
+}
+
+func (impl *canonicalChainBuilderImpl) Tip() *types.Header {
+	return impl.tip.header
+}
+
+func (impl *canonicalChainBuilderImpl) Headers() []*types.Header {
+	var headers []*types.Header
+	node := impl.tip
+	for node != nil {
+		headers = append(headers, node.header)
+		node = node.parent
+	}
+	reverse(headers)
+	return headers
+}
+
+func (impl *canonicalChainBuilderImpl) HeadersInRange(start uint64, count uint64) []*types.Header {
+	headers := impl.Headers()
+	if len(headers) == 0 {
+		return nil
+	}
+	if headers[0].Number.Uint64() > start {
+		return nil
+	}
+	if headers[len(headers)-1].Number.Uint64() < start+count-1 {
+		return nil
+	}
+
+	offset := start - headers[0].Number.Uint64()
+	return headers[offset : offset+count]
+}
+
+func (impl *canonicalChainBuilderImpl) Prune(newRootNum uint64) error {
+	if (newRootNum < impl.root.header.Number.Uint64()) || (newRootNum > impl.Tip().Number.Uint64()) {
+		return errors.New("canonicalChainBuilderImpl.Prune: newRootNum outside of the canonical chain")
+	}
+
+	newRoot := impl.tip
+	for newRoot.header.Number.Uint64() > newRootNum {
+		newRoot = newRoot.parent
+	}
+
+	impl.root = newRoot
+	return nil
+}
+
+func mapSlice[T any, U any](s []T, mapFunc func(T) U) []U {
```

Comment:
should we move this to `./common/map` package?

Reply:
There's also `reverse` - https://github.com/ledgerwatch/erigon/blob/devel/polygon/sync/heimdall.go#L64 . I'm used to having such collection helpers globally. Maybe move it to erigon-lib/common/collections.go ?